### DISCUSSION TITLE
Merges to the release branch decreases the value of CommitsSinceVersionSource

### DISCUSSION THREAD

**Discussion by AndreasFetzer** [GitTools/GitVersion#3211] [Q&A]
Hi,

we have observed a strange behaviour in our repository: The value of CommitsSinceVersionSource changes with every merge commit of our remote release branch to the local one.

Some background information:
* We have the following branches:
    * master (is basically the develop)
    * release-VERSION (release branch and stable)
    * occasionally feature branches
* Our workflow is the following:
    * Outside of a release phase we are working on the master
    * When entering a release we create a release branch, e.g. "release-1.0.0"
    * While working on a release the bugfixes are either committed directly on the release branch or merged from a separate branch
    * Sometimes this requires the developers to merge the release branch either into their local release branch or the bugfix branch

The behaviour that we now have is:
* Suddenly after such a merge of the release branch, the value for CommitsSinceVersionSource dropped from 118 to 49
* From the diagnostic output of GitVersion I can see:

```
INFO [09/28/22 15:02:39:81] Begin: Calculating base versions
  INFO [09/28/22 15:02:39:89] Fallback base version: 0.1.0 with commit count source 13da42ccb7d53763d079f9cc98f929b1f283971c
  INFO [09/28/22 15:02:39:94] Found commit [a23128f [TICKET-61655] <bug fix>: Fix bindings] matching merge message format: Default
  INFO [09/28/22 15:02:39:94] Found commit [a23128f [TICKET-61655] <bug fix>: Fix bindings] matching merge message format: Default
  INFO [09/28/22 15:02:39:94] Found commit [a23128f [TICKET-61655] <bug fix>: Fix bindings] matching merge message format: Default
  INFO [09/28/22 15:02:39:96] Found commit [a23128f [TICKET-61655] <bug fix>: Fix bindings] matching merge message format: Default
  INFO [09/28/22 15:02:39:96] Found commit [a23128f [TICKET-61655] <bug fix>: Fix bindings] matching merge message format: Default
  INFO [09/28/22 15:02:39:96] Merge message 'Merge branch 'release-1.3.0' of our-repo into release-1.3.0': 1.3.0 with commit count source 44973ff567c372b23b62c76a81f533de10b1d121
  INFO [09/28/22 15:02:39:97] Merge message 'Merge branch 'release-1.3.0' of our-repo into release-1.3.0': 1.3.0 with commit count source 0fb47744c5dbc428bffd47a80e80f0582cd39819
  INFO [09/28/22 15:02:39:97] Merge message 'Merge branch 'release-1.3.0' of our-repo into release-1.3.0': 1.3.0 with commit count source 8a3113e63a7235a98c0acc9164acdd548922b9dc
  INFO [09/28/22 15:02:39:97] Merge message 'Merge branch 'release-1.3.0' of our-repo into release-1.3.0': 1.3.0 with commit count source bbd0598f86c6046dd4b3d417e47822d1b64a1eed
  INFO [09/28/22 15:02:39:97] Merge message 'Merge branch 'release-1.3.0' of our-repo into release-1.3.0': 1.3.0 with commit count source 50ca6828a597945f24f4dd8306c5dd54cdbf528b
    ...
    ...
  INFO [09/28/22 15:02:40:03] Begin: Finding branch source of 'release-1.3.0'
    ...
    ...
  INFO [09/28/22 15:02:40:12] End: Finding branch source of 'release-1.3.0' (Took: 92.01ms)
  INFO [09/28/22 15:02:40:12] Version in branch name: 1.3.0 with commit count source a23128ff788b944b3f7591be4af70859ba134313
  INFO [09/28/22 15:02:40:23] Found multiple base versions which will produce the same SemVer (1.3.0), taking oldest source for commit counting (Merge message 'Merge branch 'release-1.3.0' of our-repo into release-1.3.0')
  INFO [09/28/22 15:02:40:23] Base version used: Merge message 'Merge branch 'release-1.3.0' of our-repo into release-1.3.0': 1.3.0 with commit count source 50ca6828a597945f24f4dd8306c5dd54cdbf528b
  INFO [09/28/22 15:02:40:23] End: Calculating base versions (Took: 422.22ms)
  INFO [09/28/22 15:02:40:23] 49 commits found between 50ca682 Merge branch 'release-1.3.0' of our-repo into release-1.3.0 and a23128f [TICKET-61655] <bug fix>: Fix bindings
  ...

```
So GitVersion is finding **THE LAST 5** merge commits of the release branch and taking the oldest one as source for the commit counting.
Now if I reset the release branch to the commit before the latest merge commit, it will again find the last 5 merges (but now the oldest one is different) and the CommitsSinceVersionSource value will change again.

I searched quiet a while for an explanation but I was not able to find an answer. And I was also not able to reproduce this with a simple test repository. Even if I merged the remote release branch several times in my test repository, GitVersion takes a different source for commit counting:

`INFO [09/28/22 15:14:37:17] Found multiple base versions which will produce the same SemVer (1.0.0), taking oldest source for commit counting (Version in branch name)`

Can someone explain me this behaviour of GitVersion? Are we doing something wrong?

I would like to avoid this issue in future releases.

Thanks in advance and best regards,
Andreas

**Comment by asbjornu** [comment]
I think you may be hit by #3041. Please read it and the referenced issues and let me know if it explains the behavior you're seeing.
